src/network/connection_handle.cpp

```diff
@@ -0,0 +1,627 @@
+        return result;
+      }
+    }
+  }
+  return WriteState::WRITE_COMPLETE;
+}
+
+Transition ConnectionHandle::CloseSocket() {
+  LOG_DEBUG("Attempt to close the connection %d", sock_fd_);
+  // Remove listening event
+  handler_->UnregisterEvent(network_event);
+  handler_->UnregisterEvent(workpool_event);
+  for (;;) {
+    int status = close(sock_fd_);
+    if (status < 0) {
+      // failed close
+      if (errno == EINTR) {
+        // interrupted, try closing again
+        continue;
+      }
+    }
+    LOG_DEBUG("Already Closed the connection %d", sock_fd_);
+    return Transition::NONE;
+  }
+}
+
+Transition ConnectionHandle::Wait() {
+  // TODO(tianyu): Maybe we don't need this state? Also, this name is terrible
+  UpdateEventFlags(EV_READ | EV_PERSIST);
+  return Transition::PROCEED;
+}
+
+Transition ConnectionHandle::Process() {
+  // TODO(tianyu): further simplify this logic
+  if (protocol_handler_ == nullptr) {
+    // Process initial
+    if (ssl_sent_) {
+      // start SSL handshake
+      // TODO: consider free conn_SSL_context
+      conn_SSL_context = SSL_new(PelotonServer::ssl_context);
+      if (SSL_set_fd(conn_SSL_context, sock_fd_) == 0) {
+        LOG_ERROR("Failed to set SSL fd");
+        PL_ASSERT(false);
+      }
+      int ssl_accept_ret;
+      if ((ssl_accept_ret = SSL_accept(conn_SSL_context)) <= 0) {
+        LOG_ERROR("Failed to accept (handshake) client SSL context.");
+        throw NetworkProcessException("ssl error: " + std::to_string(SSL_get_error(conn_SSL_context, ssl_accept_ret)));
+      }
+      LOG_ERROR("SSL handshake completed");
+      ssl_sent_ = false;
+    }
+
+    switch (ProcessInitial()) {
+      // TODO(tianyu): Should convert to use Transition in the top level method
+      case ProcessResult::COMPLETE:
+        return Transition::PROCEED;
+      case ProcessResult::MORE_DATA_REQUIRED:
+        return Transition::NEED_DATA;
+      default:
+        throw NetworkProcessException("Unexpected Process Initial result");
+    }
+  } else {
+    ProcessResult status = protocol_handler_->Process(*rbuf_, (size_t) handler_->Id());
+
+    switch (status) {
+      case ProcessResult::MORE_DATA_REQUIRED:
+        return Transition::NEED_DATA;
+      case ProcessResult::COMPLETE:
+        return Transition::PROCEED;
+      case ProcessResult::PROCESSING:
+        EventUtil::EventDel(network_event);
+        LOG_TRACE("ProcessResult: queueing");
+        return Transition::GET_RESULT;
+      case ProcessResult::TERMINATE:
+        return Transition::FINISH;
+    }
+  }
+  throw NetworkProcessException("Unexpected process result");
+}
+
+Transition ConnectionHandle::ProcessWrite() {
+  // TODO(tianyu): Should convert to use Transition in the top level method
+  switch (WritePackets()) {
+    case WriteState::WRITE_COMPLETE:
+      UpdateEventFlags(EV_READ | EV_PERSIST);
+      return Transition::PROCEED;
+    case WriteState::WRITE_NOT_READY:
+      return Transition::NONE;
+  }
```

Comment:
default?

Reply:
This should go directly to the next line which throws an exception.

This is here because gcc on linux bitches about uncaught states event though there are only two symbols for WriteState.